libcxx/include/__memory/tombstone_traits.h

```diff
@@ -0,0 +1,288 @@
+
+template <class _Tp, class _Payload, bool = __tombstone_traits<_Tp>::__is_disengaged_offset_ == 0>
+struct __tombstone_is_disengaged {
+  using _TombstoneLayout = __tombstone_traits<_Tp>;
+  using _IsDisengagedT   = remove_cv_t<decltype(_TombstoneLayout::__disengaged_value_)>;
+
+  char __padding_[_TombstoneLayout::__is_disengaged_offset_];
+  _IsDisengagedT __is_disengaged_;
+};
+
+template <class _Tp, class _Payload>
+struct __tombstone_is_disengaged<_Tp, _Payload, true> {
+  using _TombstoneLayout = __tombstone_traits<_Tp>;
+  using _IsDisengagedT   = remove_cv_t<decltype(_TombstoneLayout::__disengaged_value_)>;
+
+  _IsDisengagedT __is_disengaged_;
+};
+
+template <class _Tp, class _Payload, bool = __tombstone_traits<_Tp>::__is_disengaged_offset_ == 0>
+struct __tombstone_data {
+  using _TombstoneLayout = __tombstone_traits<_Tp>;
+  using _IsDisengagedT   = remove_cv_t<decltype(_TombstoneLayout::__disengaged_value_)>;
+
+  static_assert(is_trivial<_IsDisengagedT>::value, "disengaged type has to be trivial!");
+  static_assert(_TombstoneLayout::__is_disengaged_offset_ >= __datasizeof_v<_Payload>);
+
+  _LIBCPP_NO_UNIQUE_ADDRESS _Payload __payload_;
+  char __padding_[_TombstoneLayout::__is_disengaged_offset_ - __datasizeof_v<_Payload>];
+  _IsDisengagedT __is_disengaged_;
+
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI constexpr __tombstone_data(_Args&&... __args)
+      : __payload_(std::forward<_Args>(__args)...), __is_disengaged_(_TombstoneLayout::__disengaged_value_) {}
+};
+
+template <class _Tp, class _Payload>
+struct __tombstone_data<_Tp, _Payload, true> {
+  using _TombstoneLayout = __tombstone_traits<_Tp>;
+  using _IsDisengagedT   = remove_cv_t<decltype(_TombstoneLayout::__disengaged_value_)>;
+
+  _IsDisengagedT __is_disengaged_;
+  _LIBCPP_NO_UNIQUE_ADDRESS _Payload __payload_;
+
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI constexpr __tombstone_data(_Args&&... __args)
+      : __is_disengaged_(_TombstoneLayout::__disengaged_value_), __payload_(std::forward<_Args>(__args)...) {}
+};
+
+template <class _Tp, class _Payload, bool = is_trivially_destructible_v<_Tp> && is_trivially_destructible_v<_Payload>>
+union _MaybeTombstone {
+  using _TombstoneLayout = __tombstone_traits<_Tp>;
+  using _TombstoneData   = __tombstone_data<_Tp, _Payload>;
+
+  _Tp __value_;
+  _TombstoneData __tombstone_;
+
+  template <class... _Args>
+  constexpr _MaybeTombstone(__init_disengaged_t, _Args&&... __args) : __tombstone_(std::forward<_Args>(__args)...) {}
+
+  template <class... _Args>
+  constexpr _MaybeTombstone(__init_engaged_t, _Args&&... __args) : __value_(std::forward<_Args>(__args)...) {}
+
+  _MaybeTombstone(const _MaybeTombstone&)            = default;
+  _MaybeTombstone(_MaybeTombstone&&)                 = default;
+  _MaybeTombstone& operator=(const _MaybeTombstone&) = default;
+  _MaybeTombstone& operator=(_MaybeTombstone&&)      = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr bool __is_engaged() const noexcept {
```

Comment:
I understand it might not actually be possible to implement it that way, but I would find an interface like this for `tombstone_traits` to be more intuitive:

```c++
// Option #1
template <>
struct __tombstone_traits<std::string> {
  static constexpr bool __is_disengaged(std::byte (&__bytes)[sizeof(std::string)]) {
    bool __is_short = __bytes[__builtin_offsetof(std::string, something)] == 0;
    std::size_t __size = ...;
    return __is_short && __size > __min_cap;
  }
};


// Option #2
template <class _Tp, std::size_t _Offset, auto _DisengagedValue>
struct __tombstone_traits_from_offset {
  static constexpr auto __disengaged_value_    = _DisengagedValue;
  static constexpr size_t __is_disengaged_offset_ = _Offset;

  static constexpr bool __is_disengaged(std::byte (&__bytes)[sizeof(_Tp)]) {
    __tombstone_is_disengaged<_Tp, _Payload> __is_disengaged;
    static_assert(sizeof(__tombstone_is_disengaged<_Tp, _Payload>) <= sizeof(_MaybeTombstone));
    __builtin_memcpy(&__is_disengaged, this, sizeof(__tombstone_is_disengaged<_Tp, _Payload>));
    return __is_disengaged.__is_disengaged_ != _TombstoneLayout::__disengaged_value_;
  }
};

template <>
struct __tombstone_traits<std::string> : __tombstone_traits_from_offset<std::string, 0, std::uint8_t(65)> { };
```

I think what I'm after is to avoid basing the whole tombstone mechanism on passing a magic value and a magic offset, and instead use something where (in theory) arbitrary code could run. In practice, I understand that we're limited in what we can do since there may not be an actual object in memory, but it still seems like defining `__is_disengaged` in terms of a function (with potentially a commonly-used helper) is more flexible. This would allow for example checking more than just integral equality to determine whether the object is engaged or not.

Reply:
I played around with this some more:

```c++
#define private public
#include <string>
#include <cassert>

template <class _Tp>
struct __tombstone_traits;

template <>
struct __tombstone_traits<std::string> {
  static bool __is_disengaged(char const (&__bytes)[sizeof(std::string)]) {
    std::string::__rep __representation;
    __builtin_memcpy(&__representation, __bytes, sizeof(__representation));
    return !__representation.__s.__is_long_ && __representation.__s.__size_ > std::string::__min_cap;
  }

  static void __set_disengaged(char (&__bytes)[sizeof(std::string)]) {
    std::string::__rep __representation;
    __builtin_memcpy(&__representation, __bytes, sizeof(__representation));
    __representation.__s.__is_long_ = false;
    __representation.__s.__size_    = std::string::__min_cap + 1;
    __builtin_memcpy(__bytes, &__representation, sizeof(__representation));
  }
};

template <class _Tp>
struct Optional {
  union {
    _Tp value;
    char bytes[sizeof(_Tp)];
  };

  constexpr Optional() : bytes{} {
    if (!std::is_constant_evaluated()) {
      __tombstone_traits<_Tp>::__set_disengaged(bytes);
    }
  }

  constexpr explicit Optional(_Tp const& val) : value(val) {}

  constexpr bool is_engaged() const {
    if (std::is_constant_evaluated()) {
      return !__builtin_constant_p(bytes[0]);
    } else {
      return !__tombstone_traits<_Tp>::__is_disengaged(bytes);
    }
  }

  constexpr _Tp& operator*() {
    assert(is_engaged());
    return value;
  }

  constexpr ~Optional() {
    if (is_engaged()) {
      std::destroy_at(std::addressof(value));
    }
  }
};

constexpr bool test() {
  {
    Optional<std::string> opt;
    assert(!opt.is_engaged());
  }

  // Short string
  {
    Optional<std::string> opt("short");
    assert(opt.is_engaged());
    assert(*opt == "short");
  }

  // Long string
  {
    Optional<std::string> opt(
        "looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong");
    assert(opt.is_engaged());
    assert(*opt == "looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong");
  }

  return true;
}

int main(int, char**) {
  test();
  static_assert(test());

  return 0;
}
```

Godbolt: https://godbolt.org/z/cfqev914a